### PR TITLE
github_packages: Add sh.brew.bottle.glibc.version

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Add `glibc_version` to `os.version` so that `brew install` can verify that the host has sufficiently recent version of glibc, and install the brewed glibc if not.

### CODE REVIEW

File: Library/Homebrew/github_packages.rb [hunk 1]
```diff
@@ -235,10 +235,18 @@ def upload_bottle(user, token, skopeo, formula_full_name, bottle_hash, dry_run:)
...
+      os_version = tab["built_on"]["os_version"] if tab["built_on"].present?
+      os_version = case os
+      when "darwin"
+        os_version || "macOS #{MacOS::Version.from_symbol(bottle_tag)}"
+      when "linux"
+        glibc_version = tab["built_on"]["glibc_version"] if tab["built_on"].present?
+        os_version ||= "Ubuntu 16.04.7"
+        glibc_version ||= "2.23"
+        os_version = os_version.delete_suffix " LTS"
+        "#{os_version} glibc #{glibc_version}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:247]: This isn't part of the OS version. I'd support it being added as additional metadata.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by MikeMcQuaid** [comment]
> Add `glibc_version` to `os.version` so that `brew install` can verify that the host has sufficiently recent version of glibc, and install the brewed glibc if not.

`brew install` won't query `os.version` here. If it's specific to GitHub Packages I'd suggest it be part of https://github.com/Homebrew/brew/pull/10983 and it'll be significantly easier to parse if it's a separate annotation rather than something that needs extracted from the `os.version`.

**Comment by sjackman** [comment]
>> Add `glibc_version` to `os.version` so that `brew install` can verify that the host has sufficiently recent version of glibc, and install the brewed glibc if not.

> `brew install` won't query `os.version` here. If it's specific to GitHub Packages I'd suggest it be part of https://github.com/Homebrew/brew/pull/10983 and it'll be significantly easier to parse if it's a separate annotation rather than something that needs extracted from the `os.version`.

https://github.com/Homebrew/brew/pull/10983/files#diff-4890b590cb90a2c9ef5af6c48b020b952b712a1b080fa39be1c377de8e8a7c62R75

My PR to install bottles from GitHub Packages already supports querying and selecting the correct bottle based on `os.version` for macOS. I'd like to add support for that feature to Linux as well, where the version of `glibc` is what matters. Parsing is straight forward with `os_version[/glibc ([0-9.]+)/, 1]`

**Comment by sjackman** [comment]
> If it's specific to GitHub Packages I'd suggest it be part of #10983

My reason to include it in this PR rather than #10983, is that I want the metadata in before bulk uploading Brewsci/bio bottles to GHCR. My PR #10983 to install from GHCR and consume this metadata is not yet ready to be merged, but this PR is to create the metadata, so that it can be consumed.

**Comment by sjackman** [comment]
This year we're planning on migrating building bottles for Linux from Ubuntu 16.04 to Ubuntu 18.04. We need this metadata to ensure that only compatible bottles are installed by the client and ensure that this migration goes smoothly.

**Comment by MikeMcQuaid** [comment]
> I'd like to add support for that feature to Linux as well, where the version of `glibc` is what matters. Parsing is straight forward with `os_version[/glibc ([0-9.]+)/, 1]`

It would be more straightforward still when you don't need to filter based on a subset of the OS version. It also isn't an OS version which is a metadata cleanliness argument for me.

> This year we're planning on migrating building bottles for Linux from Ubuntu 16.04 to Ubuntu 18.04. We need this metadata to ensure that only compatible bottles are installed by the client and ensure that this migration goes smoothly.

Cool. Fine to add it in some form but let's not put it in the OS version when it's not the OS version.

**Review by MikeMcQuaid — DISMISSED** [review]
In the interest of compromise on this: my comments still stand but I'll let you decide what you think is best and merge that.

File: Library/Homebrew/github_packages.rb (hunk 1 above)
Comment (sjackman) [reply Library/Homebrew/github_packages.rb]: I've moved it to the annotation `sh.brew.bottle.glibc.version`.

File: Library/Homebrew/github_packages.rb [hunk 2]
```diff
@@ -235,10 +232,18 @@ def upload_bottle(user, token, skopeo, formula_full_name, bottle_hash, dry_run:)
...
+        os_version || "macOS #{MacOS::Version.from_symbol(bottle_tag)}"
+      when "linux"
+        (os_version || "Ubuntu 16.04.7").delete_suffix " LTS"
       else
-        MacOS::Version.from_symbol(bottle_tag).to_s
+        os_version
+      end
+
+      glibc_version = if os == "linux"
+        (tab["built_on"]["glibc_version"] if tab["built_on"].present?) || "2.23"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:246]: @sjackman Can this `2.23` here be cleaned up to use a constant or a value from a formula instead?

File: Library/Homebrew/github_packages.rb [hunk 3]
```diff
@@ -235,10 +232,18 @@ def upload_bottle(user, token, skopeo, formula_full_name, bottle_hash, dry_run:)
...
       raise TypeError, "unknown tab['built_on']['os']: #{tab["built_on"]["os"]}" if os.blank?
 
-      os_version = if tab["built_on"].present? && tab["built_on"]["os_version"].present?
-        tab["built_on"]["os_version"]
+      os_version = tab["built_on"]["os_version"] if tab["built_on"].present?
+      os_version = case os
+      when "darwin"
+        os_version || "macOS #{MacOS::Version.from_symbol(bottle_tag)}"
+      when "linux"
+        (os_version || "Ubuntu 16.04.7").delete_suffix " LTS"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:240]: @sjackman Can this `Ubuntu 16.04.7` here be cleaned up to use a constant instead?

File: Library/Homebrew/github_packages.rb [hunk 4]
```diff
@@ -205,10 +205,7 @@ def upload_bottle(user, token, skopeo, formula_full_name, bottle_hash, dry_run:)
       "org.opencontainers.image.url"           => bottle_hash["formula"]["homepage"],
       "org.opencontainers.image.vendor"        => org,
       "org.opencontainers.image.version"       => version,
-    }
-    formula_annotations_hash.each do |key, value|
-      formula_annotations_hash.delete(key) if value.blank?
-    end
+    }.compact
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:208]: This behaviour change is a bit undesirable as it'll allow empty strings whereas the previous version didn't.

**Review by MikeMcQuaid — COMMENTED** [review]
Some post-merge cleanup requests

File: Library/Homebrew/github_packages.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/github_packages.rb:240]: @sjackman Also the `.7` here definitely seems too specific.

File: Library/Homebrew/github_packages.rb (hunk 4 above)
Comment (sjackman) [reply Library/Homebrew/github_packages.rb:208]: Ah, I missed that distinction. I'll open a PR to fix it.

File: Library/Homebrew/github_packages.rb (hunk 3 above)
Comment (sjackman) [reply Library/Homebrew/github_packages.rb:240]: Opened PR https://github.com/Homebrew/brew/pull/11035

File: Library/Homebrew/github_packages.rb (hunk 4 above)
Comment (sjackman) [reply Library/Homebrew/github_packages.rb:208]: I see you already did. I opened this PR https://github.com/Homebrew/brew/pull/11036 with a stylistic improvement IMO to use `reject v.blank?` to remove empty hash values.

**Pull request merged**